internal/util-complete/src/main/scala/sbt/internal/util/LineReader.scala

```diff
@@ -191,11 +182,50 @@ trait LineReader {
   def readLine(prompt: String, mask: Option[Char] = None): Option[String]
 }
 
-final class FullReader(
+object LineReader {
+  import scala.util.Properties
+  val isEmacsShell = (
+    (Properties.propOrElse("env.emacs", "") != "") || // same as scala.tools.nsc.Properties.isEmacsShell
+      Properties.envOrNone("INSIDE_EMACS").isDefined // preferred by emacs
+  )
+  val haveInteractiveConsole = !isEmacsShell
+  val HandleCONT =
+    !java.lang.Boolean.getBoolean("sbt.disable.cont") && Signals.supported(Signals.CONT)
+
+  def simple(
+      historyPath: Option[File],
+      handleCONT: Boolean,
+      injectThreadSleep: Boolean
+  ): LineReader =
+    if (haveInteractiveConsole)
+      new SimpleReaderJLineImpl(historyPath, handleCONT, injectThreadSleep)
+    else new NonInteractiveConsoleImpl()
+
+  def simple: LineReader = simple(None, HandleCONT, false)
+
+  def full(
+      historyPath: Option[File],
+      complete: Parser[_],
+      handleCONT: Boolean = HandleCONT,
+      injectThreadSleep: Boolean = false
+  ): LineReader =
+    if (haveInteractiveConsole)
+      new FullReaderJLineImpl(historyPath, complete, handleCONT, injectThreadSleep)
+    else new NonInteractiveConsoleImpl()
```

Comment:
The returned LineReader is expected to be interruptible when `injectThreadSleep` is `true`.

Reply:
Ok, I'll have a look

Reply:
Eugene, why do you use new `FileInputStream(FileDescriptor.in)` instead of `System.in`? Is it prooven to have some difference?

Reply:
I am not sure. I think I copied whatever that was before it.